### PR TITLE
TEAMNADO-1991 Render Blank State for Manifest page

### PR DESCRIPTION
**Pull request by joeldmyers**
### Changes
- Page renders with explanatory text component per wireframe, when no manifests found and org_admin status is true
- Page renders with no permissions component, when no manifests found and org_admin status is false
- Authentication wrapper component added.
- Light refactor of `<Processing />` component to facilitate reuse.
- Extracted `authenticateUser` method from useSatelliteManifests hooks into utilities/platformServices.ts to facilitate reuse.

Note: because this story affected the loading status, as well as the "empty state" of the Satellite Manifest Panel, the loading logic was removed from the SatelliteManifestPanel itself, to avoid a UI scenario where a user sees the Manifest Panel table only to a) see the Create Manifest blank state, or b) see the No Permissions blank state.

### CODE REVIEW

File: src/components/emptyState/index.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import NoResults from './NoResults';
+import NoSatelliteManifests from './NoSatelliteManifests';
 import Processing from './Processing';
 
-export { NoResults, Processing };
+export { NoResults, NoSatelliteManifests, Processing };
```
Comment (zanewoodfin) [inline src/components/emptyState/index.js:5]: We may want to think about relocating and/or renaming some of these empty states because the distinction between `NoResults` and `NoSatelliteManifests` is not immediately apparent.

File: src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx [hunk 2]
```diff
@@ -39,22 +52,113 @@ describe('Satellite Manifests Page', () => {
...
+
+    (authenticateUser as jest.Mock).mockReturnValue(
+      Promise.resolve({
+        identity: {
+          user: {
+            is_org_admin: true
+          }
+        }
+      })
+    );
```
Comment (zanewoodfin) [inline src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx:68]: Since you are using this same mock repeatedly, you should pull it out into a before block for the whole outer describe block and assign an `isOrgAdmin` variable to `is_org_admin`.  Then add inner describe blocks for each of the states of `isOrgAdmin` in which you assign the applicable value to it.

File: src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx [hunk 3]
```diff
@@ -39,22 +52,113 @@ describe('Satellite Manifests Page', () => {
...
-      </Provider>
+      <Authentication>
+        <Provider store={init().getStore()}>
+          <Router>
+            <QueryClientProvider client={queryClient}>
+              <SatelliteManifestPage />
+            </QueryClientProvider>
+          </Router>
+        </Provider>
+      </Authentication>
```
Comment (zanewoodfin) [inline src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx:84]: Can you pull this whole thing being rendered out into a variable or something, so you don't have to repeat the same stuff for each test?

File: src/utilities/__tests__/platformServices.test.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import { authenticateUser } from '../platformServices';
+
+describe('Authenticate User method', () => {
+  it('should return a promise with user data', () => {
+    window.insights = {
+      chrome: {
+        auth: {
+          getUser: jest.fn().mockImplementation(
+            () =>
```
Comment (zanewoodfin) [inline src/utilities/__tests__/platformServices.test.ts:9]: Seems like all the code around this function could be happening just once for the whole file in a before block, with just this function being defined independently for each test.

**Review by zanewoodfin — CHANGES_REQUESTED** [review]

File: src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx [hunk 5]
```diff
@@ -1,28 +1,42 @@
...
       <Main>
         <>
-          {!error && <SatelliteManifestPanel isLoading={isLoading} data={data} />}
+          {isLoading && <Processing />}
+
+          {!isLoading && data?.length > 0 && <SatelliteManifestPanel data={data} />}
+
+          {!isLoading && data?.length === 0 && user.isOrgAdmin === true && <NoSatelliteManifests />}
+
+          {!isLoading && data?.length === 0 && user.isOrgAdmin === false && (
```
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx]: Aha good call; I'd had to add that to the NoSatelliteManifests component check due to a bug where it shows when it's not supposed to, but it appears unnecessary for the NotAuthorized component

File: src/components/emptyState/NoSatelliteManifests.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    <EmptyState variant={EmptyStateVariant.large}>
+      <EmptyStateIcon icon={PlusCircleIcon} />
+      <Title headingLevel="h5" size="lg">
+        Create a new manifest to export subscriptions
+      </Title>
+      <EmptyStateBody>
+        You currently have no manifests displayed. Create a new manifest to export subscriptions
+        from the Red Hat Customer Portal to your on-premise subscription management application.
+      </EmptyStateBody>
+      <Button variant="primary">Create new manifest</Button>
```
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/components/emptyState/NoSatelliteManifests.tsx:23]: So here, it is checked in the SatelliteManifestPage, in the line `{!isLoading && data?.length === 0 && user.isOrgAdmin === true && <NoSatelliteManifests />}` -- so if user.isOrgAdmin is false, this won't render (thus they can't click on the Create new manifest button), and the NotAuthorized component (with no Create button) will render instead.  

With the button above the table, it was unclear whether that should be part of this story based on this ticket.  I went ahead and added that in and updates the tests.

File: src/components/emptyState/NoSatelliteManifests.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  EmptyStateVariant,
+  Title
+} from '@patternfly/react-core';
+import { PlusCircleIcon } from '@patternfly/react-icons/dist/js/icons/plus-circle-icon';
+
+const NoSatelliteManifests: FC = () => {
+  return (
+    <EmptyState variant={EmptyStateVariant.large}>
+      <EmptyStateIcon icon={PlusCircleIcon} />
+      <Title headingLevel="h5" size="lg">
```
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/components/emptyState/NoSatelliteManifests.tsx]: Good catch; just updated.

File: src/components/SatelliteManifestPanel/SatelliteManifestPanel.tsx [hunk 8]
```diff
@@ -123,23 +119,12 @@ const SatelliteManifestPanel: FunctionComponent<SatelliteManifestPanelProps> = (
   };
 
   const count = () => {
-    return isLoading ? 0 : filteredData().length;
-  };
-
-  const emptyState = () => {
-    if (isLoading) {
-      return <Processing />;
```
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/components/SatelliteManifestPanel/SatelliteManifestPanel.tsx:131]: Good question. 

**TLDR** I think we can leave as-is and tackle the question in TEAMNADO-1940 but lmk if you disagree; I'm fine with tweaking for that case.

There's a decent chance that the current functionality in this PR covers this scenario, and nothing needs to be changed for now.  

As of right now, my plan for the create functionality is this -- the create manifest button will open a modal with a form.  Once the form is submitted, my plan is to render a spinner in the modal, post the data using useMutation from react-query, then on success, invalidate the query data for manifests (through react-query).  When this happens, isLoading is actually still false, per how react-query works, so the above would have to be modified regardless.  

React-query has a separate attribute, isFetching, that's checking if a refetch was triggered and is pending.  In this scenario, once the create manifest has succeeded, I see three options: 

1. Leave as-is with these changes, and conditionally render the spinner off of isFetching from SatelliteManifestPage, for the whole page. 
2. Leave as-is and have _no_ spinner for the refetch - React-query will mark the satellite data as stale, the table will show the old data, and a few seconds later the refetch will be done and the most up-to-date data will trigger a re-render, making it up-to-date.
3. Revert the changes I had around removing the isLoading functionality from here, but change isLoading to isFetching instead.

I'm open to reverting this back to the way it was before and making adjustments per the above for isFetching; that said, I wonder if it's ok to leave as-is for now and tackle the UI refinement in TEAMNADO-1940. 

LMK thoughts.

File: src/components/emptyState/index.js (hunk 1 above)
Comment (joeldmyers) [reply src/components/emptyState/index.js]: Yeah I actually had renamed NoResults to NoSearchResults but then I reverted back to cut down on changes/semi-related refactors.  

I think NoSearchResults (old NoResults) and CreateManifestPanel (old NoSatelliteManifests) are good names; I just pushed that change; lmk thoughts.

File: src/components/emptyState/NoSatelliteManifests.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmbuscemi) [reply src/components/emptyState/NoSatelliteManifests.tsx]: Normally, if creating isn't part of this story we would add the buttons in a future story.  Because what happens if we release this and a user is an org admin? They go to this page, see the create button and it does nothing.

I would think that https://issues.redhat.com/browse/TEAMNADO-1940 would implement the Create button and place it anywhere it's needed (e.g. on the blank state or when there is data in the table).

[the remark this answers is not in the document]
Comment (jmbuscemi) [reply src/components/emptyState/NoSatelliteManifests.tsx]: let me know if i'm off base here though because i'm pretty far out of the loop on this work

[the remark this answers is not in the document]
Comment (zanewoodfin) [reply src/components/emptyState/NoSatelliteManifests.tsx]: We already had one non-functional create button from the proof of concept.  I _think_ the idea is that we aren't releasing it till create functionality is there anyway.

File: src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx (hunk 2 above)
Comment (joeldmyers) [reply src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx]: OK, I pushed a tweak to this; I didn't do it exactly as you described -- there is a scoping issue using beforeEach with mocking the authenticateUser function, since we need to check that authenticateUser has been called for these tests.  I created a helper function, mockAuthenticateUser, that returns the authenticateUser mock; this achieves the same thing of cleaning up code repetition and is readable.  LMK thoughts.

File: src/components/emptyState/NoSatelliteManifests.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/components/emptyState/NoSatelliteManifests.tsx]: That's correct; I actually had removed it after chatting with Anand but Tramaine said she was fine with leaving the non-functional button for now.  

In any case I'll be picking up TEAMNADO-1940 here shortly and this won't be pushed to prod for a bit, so we should be ok here.

File: src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (zanewoodfin) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx]: Won't it render both the table and the not authorized component now if there are results and the person is non-admin?

My confusion wasn't really with the technical aspects of the code, but with why we would rather non-admin users see "not authorized" than just seeing a table with no results.  It seems misleading since non-admins do seem to be authorized to access the table.

File: src/components/emptyState/NoSatelliteManifests.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmbuscemi) [reply src/components/emptyState/NoSatelliteManifests.tsx]: I see...that all makes sense.  Carry on!

File: src/utilities/__tests__/platformServices.test.ts (hunk 4 above)
Comment (joeldmyers) [reply src/utilities/__tests__/platformServices.test.ts]: OK I pushed an update for that.

File: src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx]: Aha ok yes I had thought there was that scenario.  

In the ticket, https://issues.redhat.com/browse/TEAMNADO-1991, it says, per Tramaine's note, that a no access message should be presented.  I also agree that it should show a no permissions issue - if they need to use this but don't have permissions, it is probably better that they realize it's an issue with permissions, rather than an issue with no manifests existing in their account, no?  LMK if there's something I'm missing, or if we should check with Tramaine or anyone else.

File: src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx (hunk 3 above)
Comment (joeldmyers) [reply src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx]: OK Updated that.

File: src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx (hunk 2 above)
Comment (zanewoodfin) [reply src/pages/SatelliteManifestPage/__tests__/SatelliteManifestPage.test.tsx]: That's good for now.

File: src/utilities/__tests__/platformServices.test.ts (hunk 4 above)
Comment (zanewoodfin) [reply src/utilities/__tests__/platformServices.test.ts]: Could we get this to something more like the following:
```ts
let getUser;

beforeEach(() => {
  jest.clearAllMocks();
  window.insights = {
    chrome: {
      auth: {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        getUser: jest.fn().mockImplementation(
          new Promise((resolve, reject) => {
            process.nextTick(getUser(resolve, reject));
          })
        )
      }
    }
  };
});

describe('Authenticate User method', () => {
  getUser = (resolve, reject) => { reject({ error: 'Error getting user' }) };

  it('should return a promise with user data', () => {
```

Just trying to isolate the part that actually changes.

File: src/components/SatelliteManifestPanel/SatelliteManifestPanel.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (zanewoodfin) [reply src/components/SatelliteManifestPanel/SatelliteManifestPanel.tsx:131]: I personally like there to be some sort of indication that the data is updating if it's being done in response to me taking an action, so a spinner somewhere will be good.  I guess we can leave things as is and come back to it in 1940.

If the future modal closes on successful submission, I like the idea of the spinner appearing in the table as the table data reloads, just so less of the page is visibly reloading.

[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/components/SatelliteManifestPanel/SatelliteManifestPanel.tsx:131]: OK yeah I agree; I added a note to include that as part of 1940

File: src/utilities/__tests__/platformServices.test.ts (hunk 4 above)
Comment (joeldmyers) [reply src/utilities/__tests__/platformServices.test.ts]: ah ok, so I had based these off the tests in swatch for similar methods, but there is actually a Jest helper for this situation that helps clean it up a lot - jest.fn().mockResolvedValue and jest.fn().mockRejectedValue.  I had tried them before but was running into issues but got them to work this time.  Just pushed a fix and it's much cleaner now.

File: src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx:31]: OK so I made an update to show the empty table instead of NoPermissions when user is not org_admin and no manifests are found.

I noticed that the verbiage of the NoSearchResults component ("no results match the filter criteria...") didn't make sense for when a user has no manifests and hasn't been filtering.  So I created an EmptyTable component where title, body, etc. can be passed into it, and changed the language for this scenario to reflect that no manifests were found under the account.

Abstracting this out may be a bit overkill and we could just have a <NoSearchResults /> and a <NoManifestsFound /> component, if we don't anticipate reusing this a lot in the future.

LMK thoughts.

[the remark this answers is not in the document]
Comment (zanewoodfin) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx:31]: I was thinking about that same issue of the no search results text not making much sense, but decided not to worry about it now because this story seems to be expanding a bit and this is something we would normally be provided with content guidance on.

The EmptyTable component may be a bit much.  Even with just these two states, they don't have the same needs.  I don't think the state representing there being no manifests at all necessarily even needs a button, and another state may have multiple buttons or an external link, etc.

File: src/components/emptyState/index.js (hunk 1 above)
Comment (zanewoodfin) [reply src/components/emptyState/index.js]: I think they could use more rework in the future, but this accomplishes differentiating them better now.

File: src/utilities/__tests__/platformServices.test.ts (hunk 4 above)
Comment (zanewoodfin) [reply src/utilities/__tests__/platformServices.test.ts]: Yes, I like these other mocks more.  Thanks!

File: src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeldmyers) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx:31]: OK sounds good; I just went ahead and went back to having two separate components, NoSearchResults and NoManifestsFound.  I removed the button from NoManifestsFound as well.

[the remark this answers is not in the document]
Comment (zanewoodfin) [reply src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx:31]: Okay.  Seems good for now.

**Review by zanewoodfin — APPROVED** [review]
